src/chatd.cpp

```diff
@@ -4315,11 +4471,116 @@ void FilteredHistory::clear()
     init();
 }
 
+HistSource FilteredHistory::loadHistory(uint32_t count)
+{
+    uint32_t messagesLoaded = 0;
+    if ((mOldestNotifyMsg != --mBuffer.end() && mBuffer.size() > 1) || mFirstNotification)
+    {
+        mFirstNotification = false;
+        for (auto it = mOldestNotifyMsg; it != mBuffer.end() && messagesLoaded < count; it++)
+        {
+            Idx index = mNewest - std::distance(mBuffer.begin(), it);
+            CALL_LISTENER_FH(onLoaded,  &(*(*it)), index);
+            messagesLoaded++;
+            mOldestNotifyMsg = it;
+        }
+
+        if (messagesLoaded)
+        {
+            return HistSource::kHistSourceRam;
+        }
+    }
+
+    uint32_t messagesLoadedFromdb = loadHistoryFromDb(count);
+    mOldestNotifyMsg = --mBuffer.end();
+    if (messagesLoadedFromdb)
+    {
+        mFirstNotification = false;
+        return HistSource::kHistSourceDb;
+    }
+
+    if (!hasAllHistory() && mChat->connection().isOnline())
+    {
+        mFetchingFromServer = true;
+        mChat->requestNodeHistoryFromServer(count);
```

Comment:
This line breaks the abstraction principle of the `FilteredHistory` class. It was designed to support different types of history, which would be very useful in the future to load different kind of messages, rather than node-attachments. If we want to achieve that goal, we should somehow instantiate objects of `FilteredHistory` by specifying the type of filter.
Let's discuss about it tomorrow in the office :)